src/arrayMethodJoin.js

```diff
@@ -5,7 +5,36 @@
  */
 function applyCustomJoin() {
   [].__proto__.join2 = function(separator) {
-    // write code here
+    let newProto = '';
+    let item = '';
+
+    if ((separator !== undefined) && (this.length === 0)) {
+      return newProto;
+    }
+
+    if (separator !== undefined) {
+      for (let i = 0; i <= this.length - 2; i++) {
+        if ((this[i] === null)
+         || (this[i] === undefined)) {
+          item = '';
+        } else {
+          item = this[i];
+        };
+        newProto += `${item}` + separator;
+      }
+
+      if ((this[this.length - 1] !== null)
+       && (this[this.length - 1] !== undefined)) {
+        newProto += `${this[this.length - 1]}`;
+      }
+    } else {
+      for (let i = 0; i <= this.length - 2; i++) {
+        newProto += `${this[i]}` + ',';
+      }
+      newProto += `${this[this.length - 1]}`;
+    }
```

Comment:
youe algorithm is very difficult, let's simplify it
actually you need just one loop where you will have two checks:
1. if your current element isn't null and isn't undefined, you will add it to the resulted string
2. if your element isn't last, you will add separator to the resulted string

that's it